tryoutmanager/models.py

```diff
@@ -0,0 +1,139 @@
+                               Column('submission_id',
+                                      UUID,
+                                      ForeignKey('submissions.id')))
+
+
+# Tables corresponding to concrete models
+class AdminUser(Base):
+    __tablename__ = 'admin_users'
+
+    email = Column(String, nullable=False, unique=True, primary_key=True)
+    password = Column(String, nullable=False)
+    created_at = Column(DateTime(timezone=True), default=datetime.utcnow)
+
+
+class Marker(Base):
+    __tablename__ = 'markers'
+
+    email = Column(String, nullable=False, unique=True, primary_key=True)
+    password = Column(String, nullable=False)
+    name = Column(String, nullable=False)
+    challenges = relationship('Challenges',
+                              secondary=markers_challenges,
+                              back_populates='markers')
+    created_at = Column(DateTime(timezone=True), default=datetime.utcnow)
+
+
+class User(Base):
+    __tablename__ = 'users'
+
+    email = Column(String, primary_key=True)
+    password = Column(String, nullable=False)
+    name = Column(String, nullable=False)
+    created_at = Column(DateTime(timezone=True), default=datetime.utcnow)
+    confirmed = Column(Boolean, nullable=False, default=False)
+    email_confirmation_slug = Column(String, unique=True)
+    email_confirmation_expiry = Column(DateTime(timezone=True))
+
+
+class Challenge(Base):
+    __tablename__ = 'challenges'
+
+    id = Column(UUID, default=uuid.uuid4, primary_key=True)
+    name = Column(String, nullable=False, unique=True)
+    markers = relationship('Markers',
+                           secondary=markers_challenges,
+                           back_populates='challenges')
+    submissions = relationship('Submission',
+                               secondary=challenges_submissions,
+                               back_populates='submissions')
+    created_at = Column(DateTime(timezone=True), default=datetime.utcnow)
+
+
+class Submission(Base):
+    __tablename__ = 'submissions'
+
+    id = Column(UUID, default=uuid.uuid4, primary_key=True)
+    contestant = relationship('User', back_populates='submissions')
+    contestant_email = Column(String,
+                              ForeignKey('users.email'),
+                              nullable=False)
+    challenge = relationship('Challenge', back_populates='submissions')
+    challenge_id = Column(UUID,
+                          ForeignKey('challenges.id'),
+                          nullable=False)
+    reservation = relationship('Reservation')
+    status = Column(Enum('open',
+                         'closed',
+                         name='submission_status'),
+                    default='closed',
+                    nullable=False)
+
+    # A contestant may only have one submission per challenge
+    __table_args__ = (
+                Index('index_submissions_on_contestant_and_challenge',
+                      'contestant_email',
```

Comment:
This is good, but I think we should figure out the mechanic for reusing the same instance between 2 years.

For example:
User Nour signups in 2015
Nour completes the 2016 challenge for `algorithms`
Next year, Nour wishes to complete `algorithms` again but can't

Do we:
- Change the name of a change to include the year for each year?
     - If so, how do we hide entries from old challenges?
- Delete everything year to year? (I don't want to do this)

I'm looking for ideas.

Reply:
I hadn't considered this. Would database versioning (e.g. `pg_dump` and `DROP DATABASE` versions from year to year) be a valid approach?

Reply:
Two more approaches:

- We don't reuse challenges year-to-year. _Especially_ with returning challengers.
- We update the unique constraint to include the submission's `created_at` field.

Reply:
It's not about reusing the challenge itself but reusing the name. I'd avoid using a database dump but for now, we can build it with that in mind and make a backlog issue for later under the working title "Reuse instances for multiple tryout periods" where the description would be to solve the user story outlined above.

Reply:
That should be a fairly easy addition. Challenges already have synthetic UUID keys, so it's just a question of moving the uniqueness constraint from the challenge name to the repository directory name.

Reply:
We'll still need to figure out the pyolite aspects later. I'm sure this will arise again. But for now, I think we're solid.